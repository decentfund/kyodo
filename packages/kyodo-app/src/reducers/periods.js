import { LOAD_CURRENT_PERIOD_INFO_SUCCESS } from '../constants';

const initialState = {
  currentPeriod: {
    periodTitle: '',
    initialBalance: 0,
    currentBalance: 0,
  }
};

export default (state = initialState, action) => {
  switch (action.type) {
    case LOAD_CURRENT_PERIOD_INFO_SUCCESS:
      return { ...state, currentPeriod: { ...action.data } };
    default:
      return state;
  }
};
