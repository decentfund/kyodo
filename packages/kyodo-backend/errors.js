const SystemError = function SystemError(message) {
  this.name = 'SystemError';
  this.message = message || '';
};

SystemError.prototype = new Error();
SystemError.prototype.constructor = SystemError;

const ModelError = function ModelError(message, model, value) {
  this.name = 'ModelError';
  this.message = message || '';
  this.model = model;
  this.value = value;
};

ModelError.prototype = new Error();
ModelError.prototype.constructor = ModelError;

export { SystemError, ModelError };
