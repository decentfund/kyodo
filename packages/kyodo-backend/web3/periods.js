import Web3 from 'web3';
const provider = new Web3.providers.WebsocketProvider(
  process.env.WS_PROVIDER || 'ws://localhost:8545',
);
import TruffleContract from 'truffle-contract';
import Periods from '@kyodo/contracts/build/contracts/PeriodsV1.json';
import { getKyodo } from './kyodo';

let periods;

export const getPeriods = async () => {
  if (periods) return periods;

  const kyodo = await getKyodo();
  const periodsAddress = await kyodo.periods();
  const contract = TruffleContract(Periods);
  await contract.setProvider(provider);
  periods = await contract.at(periodsAddress);
  return periods;
};

export const getStartTime = async () => {
  const periods = await getPeriods();
  const currentPeriodStartTime = (await periods.currentPeriodStartTime()).toNumber();
  return currentPeriodStartTime;
};

export const getPeriodDaysLength = async () => {
  const periods = await getPeriods();
  const periodDaysLength = (await periods.periodDaysLength()).toNumber();
  return periodDaysLength;
};
