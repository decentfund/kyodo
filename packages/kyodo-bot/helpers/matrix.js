const isDirectChat = (room, user) => {
  let type = "room";
  // if (type === 'directMessage') {
  // state.sources[roomId].name = room.name.replace(/@(\w+):.+/, '$1');
  // }
  const allMembers = room.currentState.getMembers();
  if (
    type === "room" &&
    allMembers.length <= 2 &&
    allMembers.some(m => m.userId === user)
  ) {
    // if (
    // // allMembers.some(m => m.getDMInviter()) &&
    // ) {
    type = "directMessage";
    // }
  }

  return type === "directMessage";
};

export const getOrCreatePrivateRoom = async (client, event, alias = null) => {
  console.log("getting");
  const party = alias || event.getSender();

  let userRoom = client.getRooms().find(r => isDirectChat(r, party));
  let roomId;
  if (userRoom) roomId = userRoom.roomId;
  if (!userRoom) {
    const newChat = await client.createRoom({
      preset: "trusted_private_chat",
      invite: [party],
      is_direct: true,
    });
    roomId = newChat.room_id;
  }

  return roomId;
};
