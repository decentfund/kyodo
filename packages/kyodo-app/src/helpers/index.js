import web3 from 'web3';

export const isValidAddress = address => web3.utils.isAddress(address);
