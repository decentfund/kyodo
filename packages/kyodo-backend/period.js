import { Period, getColonyById } from './db.js';
import { dbGetAllUsers, getAllUsers } from './user.js';
import { PERIOD_TIME } from './constants/periodTime.js';

let getBalance = () => 0;
if (process.env.NODE_ENV !== 'test') {
  const { getBalance: tokenGetBalance } = require('./token');
  getBalance = tokenGetBalance;
}

let currentPeriod = 0;
console.log('PERIOD', currentPeriod);

export const initPeriod = async (blockNumber, periodId, colonyId) => {
  const colony = await getColonyById(colonyId);
  const users = await dbGetAllUsers();
  currentPeriod = periodId;
  colony.periodIds.push(currentPeriod);
  await colony.save();

  users.map(async user => {
    const balance = await getBalance(user.address, blockNumber);
    await createAndSaveNewUserPeriod({
      address: user.address,
      periodId: currentPeriod,
      balance, // current user balance
      tips: 0,
      user,
    });
  });
};

export const createAndSaveNewUserPeriod = async ({
  address,
  periodId,
  balance,
  tips,
  user,
}) => {
  let period = new Period({
    // TODO: Period title
    title: 'My new period',
    address,
    periodId,
    balance, // current user balance
    initialBalance: balance,
    user,
    tips,
  });
  await period.save();
  return period;
};

export const clearPeriods = async () => {
  return Period.deleteMany();
};

export const initiateNewPeriod = async (req, res) => {
  // verify all users are present in db
  // fetching users from smart contract

  let users = await dbGetAllUsers();
  currentPeriod++;
  users.map(async user => {
    console.log('USER INSIDE LOOP', user);
    let period = new Period({
      title: req.body.title,
      address: user.address,
      periodId: currentPeriod,
      balance: user.balance, // TODO: get the real user's balance and put it here
      initialBalance: user.balance,
      tips: 0,
      user,
    });
    await period.save();
  });
  res
    .status(200)
    .send(
      `Successfully initiated first period of a colony, let the game begin!!!`,
    );
};

export const getAllPeriods = async (req, res) => {
  let periods = await Period.find((err, periods) => {
    if (err) return console.error(err);
    res.status(200).send(periods);
  });
  return periods;
};

export const getCurrentPeriod = async (req, res) => {
  const period = await Period.find({ periodId: currentPeriod });
  res.status(200).send(period);
};

export const getCurrentPeriodTotalBalance = async (req, res) => {
  const period = await Period.aggregate([
    { $match: { periodId: currentPeriod } },
    { $group: { _id : null, balance: { $sum: "$initialBalance" } } },
    { $project: { _id: 0, balance: 1 } }
  ]);
  res.status(200).send(period[0] || {});
};

// exports.getUserPeriodBalance = async () => {};

export const getUserByAddressInPeriod = async address => {
  let user = await Period.find(
    {
      address: address,
    },
    (err, user) => {
      if (err) console.log(err);
      return user;
    },
  );
  return user;
};

export const changeUserBalance = async (address, tip) => {
  let sender = await Period.find({ address, periodId: currentPeriod });
  await Period.update(
    { address, periodId: currentPeriod },
    { $set: { balance: sender[0].balance - tip } },
    (err, res) => {
      if (err) console.log(err);
      return res;
    },
  );
};

//MEGA CRON
export const beginTheColonyCountdown = async (req, res) => {
  setTimeout(async () => {
    await initiateNewPeriod(req, res);
  }, PERIOD_TIME);
};

//TODO: startNewPeriod

export { currentPeriod };
