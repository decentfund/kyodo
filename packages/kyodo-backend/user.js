import web3 from 'web3';
import { Colony, getColonyById, getCurrentUserPeriod } from './db';
import User from './models/User';
import { createAndSaveNewUserPeriod } from './period';
const Point = require('./models/Point');
const { ModelError } = require('./errors');

export const dbAddUser = async ({
  address,
  alias,
  balance,
  domains,
  tasks,
}) => {
  const user = new User({
    address,
    alias,
    balance,
    domains,
    tasks,
    dateCreated: Date.now(),
  });

  await user.save(err => {
    if (err) return null;
  });

  return user;
};

export const addUser = async ({ alias }) => {
  const colony = await Colony.findOne();
  const user = await dbAddUser({ alias });
  await createAndSaveNewUserPeriod({
    periodId: colony.currentPeriodId,
    balance: 0,
    user,
  });
  return user;
};

export const dbGetAllUsers = async () => {
  return await User.find((err, users) => {
    if (err) {
      console.log(err);
      return [];
    }
    return users;
  });
};

export const getAllUsers = async (req, res) => {
  let users = await dbGetAllUsers();
  res.status(200).send(users);
  return users;
};

export const getUserByAddress = async (req, res) => {
  await User.find({ address: req.body.address });
};

export const getUserByAlias = async (req, res) => {
  await User.find({ alias: req.body.alias });
};

export const dbGetUserByAlias = async alias => {
  const user = await User.findOne({ alias });

  if (!user) {
    throw new ModelError('Not found', 'user', alias);
  }
  return user;
};

export const getOrCreateUser = async ({ alias }) => {
  let user;
  try {
    user = await dbGetUserByAlias(alias);
  } catch (e) {
    if (e.message === 'Not found') {
      user = await addUser({ alias });
    }
  }
  return user;
};

export const delegatePoints = async ({ amount, sender, receiver }) => {
  // Verify sender is present and has enough points
  if (!sender) throw Error('No sender specified');
  // Throw error if sender is not present
  const senderUser = await dbGetUserByAlias(sender);

  const balance = await getUserBalance(sender);

  if (balance < amount)
    throw Error(
      `Naaah, you don't have so much points. Your current point balance is ${balance}`,
    );

  // Get receiver if not present create a new one
  const receiverUser = await getOrCreateUser({ alias: receiver });

  // Create points and pass to user
  const points = new Point({
    amount,
    used: false,
    delegatee: senderUser,
    owner: receiverUser,
  });

  await points.save();

  return points;
};

export const getUserBalance = async alias => {
  const colony = await getColonyById(0);
  const currentPeriodId = colony.periodIds[colony.periodIds.length - 1];
  const { balance } = await getCurrentUserPeriod(alias, currentPeriodId);
  return balance;
};

export const setUserAlias = async ({ user, alias, blockNumber }) => {
  user.alias = alias;
  user.aliasSet = blockNumber;

  user.save(err => {
    if (err) return null;
  });

  return user;
};

export const addDomainToUser = async (req, res) => {
  // TODO:
};

export const addTaskToUser = async (req, res) => {
  // TODO:
};

export const changeUserAlias = async (req, res) => {
  // TODO:
  // this.getUserByAlias(req.body.alias);
};

export const updateUserAddress = async ({ alias, address }) => {
  if (!web3.utils.isAddress(address))
    throw Error('Address is incorrect, try again');
  const user = await dbGetUserByAlias(alias);
  if (!user) throw Error('Get some tips to get started');
  if (user.address)
    throw Error('Address is already set, I cannot change it by now');

  user.address = address;
  await user.save();
  return user;
};
