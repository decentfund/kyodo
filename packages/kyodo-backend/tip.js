import map from 'lodash/map';
import sum from 'lodash/sum';
import { Tip, Domain, getDomainByCode, Colony, getColonyById } from './db.js';
import {
  changeUserBalance,
  currentPeriod,
  getUserByAddressInPeriod,
} from './period';
import { getUserBalance, dbGetUserByAlias, getOrCreateUser } from './user';
import { dbCreateTask as createTask, getTaskByTitle } from './task';

export const sendTip = async (req, res) => {
  // TODO: colonyClient integration
  // transfer.send({ destinationAddress, amount }, options)
  // transferFrom.send({ sourceAddress, destinationAddress, amount }, options)

  let tipAmount = req.body.amount;
  if (tipAmount <= 0) {
    res.status(400).send('Seriously?!');
    return 'No money no honey, sorry';
  }

  let sender = await getUserByAddressInPeriod(req.body.from);
  let receiver = await getUserByAddressInPeriod(req.body.to);
  let senderAddress = sender[0].address;
  let senderBalance = sender[0].balance;
  let receiverAddress = receiver[0].address;

  if (!checkBalance(senderBalance, tipAmount)) {
    res.status(400).send('Inssuficient funds for the tip');
    return 'No money no honey, sorry';
  }

  await changeUserBalance(senderAddress, tipAmount);
  await changeUserBalance(receiverAddress, -tipAmount);

  let tip = new Tip({
    from: senderAddress,
    to: receiverAddress,
    amount: tipAmount,
    taskId: req.body.taskId,
    domainId: req.body.domainId,
    potId: req.body.potId,
    dateCreated: Date.now(),
    periodId: currentPeriod,
  });

  tip.save(err => {
    if (err) return console.error(err);
  });
  res
    .status(200)
    .send(
      `User ${senderAddress} Successfully tipped user ${receiverAddress} with ${tipAmount} `,
    );
};

export const sendNewTip = async ({
  sender,
  amount = 0,
  domain: code,
  title,
  receiver,
} = {}) => {
  try {
    const colony = await Colony.findOne();
    const currentPeriodId = colony.periodIds[colony.periodIds.length - 1];

    // Verify sender is specified
    if (!sender) throw Error('No sender specified');

    // Throw error if amount is not specified
    if (amount <= 0) throw new Error('No money no honey, sorry');

    // Throw error if sender is not present
    const senderUser = await dbGetUserByAlias(sender);

    // Get receiver if not present create a new one
    const receiverUser = await getOrCreateUser({ alias: receiver });

    const balance = await getUserBalance(sender);
    if (!checkBalance(balance, amount))
      throw Error(
        "You have 0 points. But don't worry! You can earn them by making contributions other people find useful. Or just by making me laugh.",
      );

    // Getting domain
    // TODO: Throw error if domain is not found
    const domain = await getDomainByCode(code);

    // TODO: Try to find task
    // FIXME: Find task in period
    // If not available create a new one
    let task = await getTaskByTitle(title);
    if (!task) {
      task = await createTask({ title });
    }

    let tip = new Tip({
      from: senderUser,
      to: receiverUser,
      amount,
      task,
      domain,
      // potId: req.body.potId,
      dateCreated: Date.now(),
      // FIXME: Get period from colony
      periodId: currentPeriodId,
    });

    await tip.save();
    return tip;
  } catch (e) {
    if (e.name === 'SystemError') {
      throw new Error(
        'Uh-oh! Something went wrong! Please contact admin from the chat and submit a ticket with details in the repo http://github.com/decentfund/kyodo',
      );
    }
    if (e.message === 'Not found') {
      if (sender === e.value) {
        throw Error('Sender is not registered');
      }
    }
    throw e;
  }
};

export const getColonyCurrentPeriodId = async () => {
  const colony = await getColonyById(0);
  const currentPeriodId = colony.periodIds[colony.periodIds.length - 1];
  return currentPeriodId;
};

export const getAllTips = async () => {
  const currentPeriodId = await getColonyCurrentPeriodId();
  let tips = await Tip.find({ periodId: currentPeriodId })
    .populate('task', 'taskTitle')
    .populate('domain', 'domainTitle')
    .populate('from')
    .populate('to');
  return tips;
};

export const getUserTips = async ({ user, periodId, direction }) => {
  const tips = await Tip.find(
    { periodId: periodId, [direction]: user._id },
    err => {
      if (err) return console.error(err);
    },
  )
    .populate('task', 'taskTitle')
    .populate('domain', 'domainTitle')
    .populate('from')
    .populate('to');

  return tips;
};

export const getAllUserTips = async ({ user, periodId = null }) => {
  const _periodId = periodId || (await getColonyCurrentPeriodId());
  const fromTips = await getUserTips({
    user,
    periodId: _periodId,
    direction: 'from',
  });
  const toTips = await getUserTips({
    user,
    periodId: _periodId,
    direction: 'to',
  });
  return [...fromTips, ...toTips];
};

export const getAllTipsInDomain = async domain => {
  const govDomain = await Domain.find({ domainTitle: domain });
  const tips = await Tip.find({ domain: govDomain });
  return {
    total: sum(map(tips, 'amount')),
    tips,
  };
};

export const checkBalance = (balance, amount) => {
  if (amount <= balance && amount > 0) return true;
  return false;
};
