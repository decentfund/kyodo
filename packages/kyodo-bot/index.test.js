import "babel-polyfill";
import moment from "moment";
import { handleBalance } from "./";
import * as matrixHelpers from "./helpers/matrix";
import * as fromUser from "@kyodo/backend/user";
import * as fromTip from "@kyodo/backend/tip";
import * as fromPeriod from "@kyodo/backend/period";
import * as fromWeb3Periods from "@kyodo/backend/web3/periods";
import { ModelError } from "@kyodo/backend/errors";

const event = {
  getSender: jest.fn(() => "@sender:matrix.org"),
};

const roomId = "r00m1d";

const originalDbGetUserByAlias = fromUser.dbGetUserByAlias;

describe("command handleBalance works properly", () => {
  // mock dbGetUserByAlias
  const user = {};
  fromUser.dbGetUserByAlias = jest.fn(() => user);

  // mock getUserTips
  const tips = [
    {
      domain: "GOV",
      amount: 10,
    },
    {
      domain: "SOCIAL",
      amount: 30,
    },
  ];
  fromTip.getUserTips = jest.fn(() => tips);

  // mock periods getStartTime
  fromWeb3Periods.getStartTime = jest.fn(() =>
    moment()
      .subtract(1, "days")
      .format("X"),
  );

  // mock periods getPeriodDaysLength
  fromWeb3Periods.getPeriodDaysLength = jest.fn(() => 45);

  // mock periods getCurrentPeriodSummary
  fromPeriod.getCurrentPeriodSummary = jest.fn(() => ({
    periodTitle: "Age",
  }));

  let client;

  beforeEach(() => {
    // mock client sendTextMessage
    const sendTextMessage = jest.fn();
    client = {
      sendTextMessage,
    };

    // mock get or create private room
    matrixHelpers.getOrCreatePrivateRoom = jest.fn(() => roomId);
  });

  it("for existing user with positive balance", async () => {
    // mock getUserBalance
    const balance = {
      balance: 10,
      initialBalance: 100,
    };
    fromUser.getUserBalance = jest.fn(() => balance);

    await handleBalance(
      event,
      {}, // passing empty room as we are mocking
      client,
    );

    expect(matrixHelpers.getOrCreatePrivateRoom.mock.calls.length).toBe(1);
    expect(client.sendTextMessage.mock.calls.length).toBe(1);
    expect(client.sendTextMessage.mock.calls[0]).toMatchSnapshot();
  });
  it("for existing user with zero balance", async () => {
    // mock getUserBalance
    const balance = {
      balance: 0,
      initialBalance: 100,
    };
    fromUser.getUserBalance = jest.fn(() => balance);

    await handleBalance(
      event,
      {}, // passing empty room as we are mocking
      client,
    );

    expect(matrixHelpers.getOrCreatePrivateRoom.mock.calls.length).toBe(1);
    expect(client.sendTextMessage.mock.calls.length).toBe(1);
    expect(client.sendTextMessage.mock.calls[0]).toMatchSnapshot();
  });
  it("for new user", async () => {
    // mock getUserBalance
    const balance = {
      balance: 0,
      initialBalance: 0,
    };
    fromUser.getUserBalance = jest.fn(() => balance);

    await handleBalance(
      event,
      {}, // passing empty room as we are mocking
      client,
    );

    expect(matrixHelpers.getOrCreatePrivateRoom.mock.calls.length).toBe(1);
    expect(client.sendTextMessage.mock.calls.length).toBe(1);
    expect(client.sendTextMessage.mock.calls[0]).toMatchSnapshot();
  });
});

describe("command handleBalance works properly for non existing user", () => {
  // mock client sendTextMessage
  const sendTextMessage = jest.fn();
  const client = {
    sendTextMessage,
  };

  beforeAll(() => {
    fromUser.dbGetUserByAlias = jest.fn(alias => {
      throw new ModelError("Not found", "user", alias);
    });
    matrixHelpers.getOrCreatePrivateRoom = jest.fn(() => roomId);
  });
  afterAll(() => {
    fromUser.dbGetUserByAlias = originalDbGetUserByAlias;
  });

  it("with positive balance", async () => {
    // mock getUserBalance
    const balance = {
      balance: 10,
      initialBalance: 100,
    };
    fromUser.getUserBalance = jest.fn(() => balance);

    await handleBalance(
      event,
      {}, // passing empty room as we are mocking
      client,
    );

    expect(matrixHelpers.getOrCreatePrivateRoom.mock.calls.length).toBe(1);
    expect(sendTextMessage.mock.calls.length).toBe(1);
    expect(sendTextMessage.mock.calls[0]).toMatchSnapshot();
  });
});
