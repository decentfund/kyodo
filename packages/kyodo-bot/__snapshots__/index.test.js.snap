// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command handleBalance works properly 1`] = `"r00m1d"`;
